### PR TITLE
[RNmobile] Add native supported VideoPress player

### LINKED ISSUE #5509: Video player
Add a native supported VideoPress video player block

### PR DESCRIPTION
**Pull request by jhnstn**
Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/5509

## Proposed changes:

Adds a minimally functioning native VideoPress player.

The changes here do not account for design and loading performance. Those updates will follow while the native block is still behind a feature flag.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:

* Go to the mobile editor
* Open a post that contains a VideoPress block.
* Verify that the block renders the video thumbnail with out cropping (try both portrait and landscape videos )
* Touch the block, note that the player does not open.
* Once the block is selected try touching the video.
* The video should play fullscreen.

* **Note**: ~There is an [open issue on Android](https://github.com/wordpress-mobile/gutenberg-mobile/issues/5565) regarding the `Sandbox` component. For now test plan will only succeed completely on iOS.~ The Sandbox issue has been resolved so testing is no longer limited to iOS. I have noticed some odd behavior when touching the video on Android but I think we'll have to fix those issues in Gutenberg and should not block these changes.

* **Note**:  The scope of the changes here are limited to existing VideoPress blocks. A block can be added via the mobile app but the post should be reloaded before verifying the test steps. Alternatively a VideoPress can be added to a post via the Web.

### CODE REVIEW

File: projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx
Comment (jhnstn) [inline projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx:110]: The web block uses [caching before passing the html](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/videopress/src/client/block-editor/blocks/video/edit.tsx#L175) to the player. 
From the hook logic it looks like the caching is only relevant when the video html changes on a mounted component, guessing that happens when changing the video. 
I'd prefer to hold on looking into this use of caching  for the native player once we add the ability to replace videos.

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx
Comment (jhnstn) [inline projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx:19]: I expect to eventually have better parity with the web `PlayerProps` type. While developing I opted for a native version of the props type. If we end up remaining divergent we can consider moving the type definition to the `types.ts` file.

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx
Comment (jhnstn) [inline projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx:36]: There is a lot more going on in the web version to assign a height to the video wrapper component. I didn't run into needing even this basic setup while testing this branch. Ideally we can keep this simple and add in complexity only if needed.

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx
Comment (jhnstn) [inline projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx:46]: I intentionally left out the caption for now. I'll add that in a following PR

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx
Comment (jhnstn) [inline projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx:9]: So I tried moving this up under the `External dependencies` block in 5f67e4aeb7a249a3ebfade9b08be9a07ea4a1098 but the commit hook is moving it back ( and the allowing an empty commit ¯\_(ツ)_/¯ )

File: projects/packages/videopress/src/client/block-editor/blocks/video/style.native.scss
Comment (SiobhyB) [inline projects/packages/videopress/src/client/block-editor/blocks/video/style.native.scss]: I believe this `TODO` could be removed now?

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx
Comment (SiobhyB) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx:9]: I recall encountering similar when working on [the `edit.native.tsx` file](https://github.com/Automattic/jetpack/blob/37b15350fb1d6abcfb3003b1241f6e4a3e4dbba9/projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx), with the workaround being to add `WordPress dependencies` above `External dependencies`. From looking at other files in the package, it seems there's a precedence for only using one of the headings or the other, too, so maybe that could be another option.

**Review by SiobhyB — COMMENTED** [review]
> Verify that the block renders the video thumbnail with out cropping

@jhnstn, this is the only part of the test plan that made me pause a little. It did seem like there was a set height that cropped longer videos in my testing:

https://user-images.githubusercontent.com/2998162/226394455-814e85ca-b016-41b9-95f2-fe09904386c1.mov

Is that expected?

Other than that, I was able to go through the testing steps as expected on iOS. As you've outlined in the PR's description, not all the functionality within the player itself looks or works exactly as expected yet, but the player itself works well.

Happy to approve once I've gained clarity on the cropping/height question. :)

**Comment by jhnstn** [comment]
> > Verify that the block renders the video thumbnail with out cropping
> 
> @jhnstn, this is the only part of the test plan that made me pause a little. It did seem like there was a set height that cropped longer videos in my testing:
> 
> Is that expected?
> 
> Other than that, I was able to go through the testing steps as expected on iOS. As you've outlined in the PR's description, not all the functionality within the player itself looks or works exactly as expected yet, but the player itself works well.
> 
> Happy to approve once I've gained clarity on the cropping/height question. :)

Thanks for reviewing @SiobhyB ! 

>  It did seem like there was a set height that cropped longer videos in my testing:

That's unfortunate. It looks like you are using the simulator right? I'm also using the simulator with the same OS and device and I'm not seeing the cropping. Also it looks like the top of the video is cropped. When I was running into cropping issues with the Sandox I noticed  the bottom was being cropped. 

Oh, I have noticed that I need to exit the editor when switching branches or making significant changes to the how the Sandbox handles the player height. Reloading metro is not enough. Could you try reloading the post from the main app page?

**Comment by SiobhyB** [comment]
@jhnstn, thanks for the pointers! I did some more testing and found that I'm only seeing the cropping after uploading a video from my device. You'll see at the `0:09`/`0:10` mark that the placeholder actually expands to the correct height, but reduces again before the video's fully uploads:

https://user-images.githubusercontent.com/2998162/226440384-d34820a9-aabf-4459-831d-57d9d6f9d411.mov

When I return back to the same post in the recording, you'll see that the video expands to the correct height. 🤔 I can confirm I'm using a simulator (iPhone 14 pro, iOS 16.2) and I uploaded to a brand new post.

Are you able to replicate the same on your side, or could this perhaps be a simulator issue? Happy to test on a physical device if you suspect this is limited to the simulator.

**Comment by jhnstn** [comment]
> I'm only seeing the cropping after uploading a video from my device.

Ah ok, that makes sense @SiobhyB. I was working on these changes while the uploading work was ongoing. The changes here are only for exiting blocks.  I tried to spell that out in the test steps ( '2. Open a post that contains a VideoPress block.' ) but I can see how that  is not super clear. I should have mentioned this has only been tested with exiting blocks and without changing the video source.

I suspect we'll need to pass some events from the VP player iframe back up to the Sandbox component to make sure the height is correct after uploading but since we are still behind a feature flag I'd prefer to get this PR merged as is and then follow up with the uploading case. 

There is quite a bit going on in the web player that handles loading the player. I opted for keeping that out of this PR to keep it simpler.

**Comment by jhnstn** [comment]
@SiobhyB  I added a new note to the PR description that I hope clarifies the limited scope of these changes.

**Comment by SiobhyB** [comment]
> The changes here are only for exiting blocks. I tried to spell that out in the test steps ( '2. Open a post that contains a VideoPress block.') but I can see how that is not super clear. I should have mentioned this has only been tested with exiting blocks and without changing the video source.

Aha, apologies for missing that @jhnstn! That makes sense, I'll go ahead to approve this PR as I can confirm that existing blocks are resized as expected :)

**Review by SiobhyB — DISMISSED** [review]
Works as outlined! 🎉

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx
Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx:9]: Yep, I've seen some files where external dependencies are under `WordPress dependencies`, so not sure what's the standard here.

Example:
https://github.com/Automattic/jetpack/blob/71b7f699487e74f23c7b322d73a25aa69aa84b8c/projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/uploader-progress.js#L1-L11

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx
Comment (fluiddot) [inline projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx]: I noticed this style (`wp-block-jetpack-videopress__video-player-overlay`) is defined in [the stylesheet of the block](https://github.com/Automattic/jetpack/blob/5cc113fe9e9d90e753b9b6098d1ccbff5527f4de/projects/packages/videopress/src/client/block-editor/blocks/video/style.native.scss). On other blocks I see we defined the styles for the specific component ([example](https://github.com/Automattic/jetpack/blob/5cc113fe9e9d90e753b9b6098d1ccbff5527f4de/projects/packages/videopress/src/client/block-editor/blocks/video/components/videopress-uploader/style.scss)), so I'm wondering if we could do the same here. This way we keep the styles close to the component where it's used.

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx
Comment (fluiddot) [inline projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx]: In React Native, you can also achieve this by passing an array ([reference](https://reactnative.dev/docs/0.66/style)):
```suggestion
		<View style={ [ style[ 'wp-block-jetpack-videopress__video-player' ], loadingStyle ] }>
```

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx
Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx:46]: I couldn't see the `Loading...` when testing. Probably this is caused by the fact that the block renders a black background ([reference](https://github.com/Automattic/jetpack/blob/5cc113fe9e9d90e753b9b6098d1ccbff5527f4de/projects/packages/videopress/src/client/block-editor/blocks/video/style.native.scss#L4)). Not sure if this background was set for testing purposes, but I'd advocate using a transparent color, WDYT?

Additionally, I noticed that for some reason videos have an extra space at the bottom. This might be expected as I can reproduce it in the web version. Not sure if it's an issue in the embedded player, but this enforces the approach of using a transparent color for the background.

Web|App
--|--
Screenshot 2023-03-21 at 12 24 24|Screenshot 2023-03-21 at 12 25 07

cc @SiobhyB

File: projects/packages/videopress/src/client/block-editor/blocks/video/style.native.scss
Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/style.native.scss]: Good point. Seems we can remove these comments.

**Review by fluiddot — COMMENTED** [review]
> The video should play fullscreen.

I noticed this behavior is only happening on iOS. On Android as far as I tested, the video is being played within the block. In fact, this leads to another issue related to not being able to play/pause the video when tapping on it (see the video capture).


https://user-images.githubusercontent.com/14905380/226595119-cbf2f63d-5afb-4e97-8045-fb07b67f6019.mp4

_Tested on Samsung Galaxy S20 FE 5G (Android 13)._

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx
Comment (jhnstn) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx:46]: > I couldn't see the `Loading...` when testing. Probably this is caused by the fact that the block renders a black background ([reference](https://github.com/Automattic/jetpack/blob/5cc113fe9e9d90e753b9b6098d1ccbff5527f4de/projects/packages/videopress/src/client/block-editor/blocks/video/style.native.scss#L4)). Not sure if this background was set for testing purposes, but I'd advocate using a transparent color, WDYT?

Ah correct, I believe that was added back when the player was just a place holder, I'll remove it.

I noticed there is some "corrective" padding in the web component but it looks like it's only used for the initial height calculation i.e. before the video is loaded. Removing the black background will help but I can look into correcting that extra spacing in a follow up PR.

**Comment by jhnstn** [comment]
> I noticed this behavior is only happening on iOS. On Android as far as I tested, the video is being played within the block. In fact, this leads to another issue related to not being able to play/pause the video when tapping on it (see the video capture).

@fluiddot , I noticed that too. I updated the note in the description about this:

> Note: ... The Sandbox issue has been resolved so testing is no longer limited to iOS. I have noticed some odd behavior when touching the video on Android but I think we'll have to fix those issues in Gutenberg and should not block these changes.

I not sure the solution is in the player component. I have a feeling it's a problem with the `WebView` on Android. It's possible we can fix it by adding passing scripts to the `Sandbox` which will happen in the `Player` component but I'd prefer to address those in a follow up PR.

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx
Comment (jhnstn) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx]: Is this preferred? The documentation first mentions:
`The style prop can be a plain old JavaScript object.`

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx
Comment (jhnstn) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx]: The Web player styles are defined in `editor.scss` ([reference](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/videopress/src/client/block-editor/blocks/video/editor.scss#L14)) . I thought there was some talk of the styles moving away from being tied to the component but I could be wrong.

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx
Comment (jhnstn) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx:9]: I opted to put the WP dependencies at the top and added an external block below. Just feels wrong to put external imports under a WP dependencies comment.
Change is in 2c936630a3cbaacf64977922c58ce383b376d49c

File: projects/packages/videopress/src/client/block-editor/blocks/video/style.native.scss
Comment (jhnstn) [reply projects/packages/videopress/src/client/block-editor/blocks/video/style.native.scss]: fixed in 2c936630a3cbaacf64977922c58ce383b376d49c

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx
Comment (jhnstn) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx:46]: fixed in 07d498f468adca0f91199ecf7eba8aac10218a0c

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx
Comment (jhnstn) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx]: fixed in 6778f08917e117cc89dbde6df309e9ccbf0e129f

**Comment by jhnstn** [comment]
@SiobhyB and @fluiddot  Thanks again for the reviews.
I know this might not be perfect but I'd like to get this merged soon to avoid more conflicts with `edit.native.tsx`. 

Siobhan, I'll need another review approval to merge if you don't mind.

Thanks again.

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx
Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx]: > The Web player styles are defined in `editor.scss` ([reference](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/videopress/src/client/block-editor/blocks/video/editor.scss#L14)) .

Ah, good point, I didn't check the styles on the web version.

> I thought there was some talk of the styles moving away from being tied to the component but I could be wrong.

That's interesting, I don't have a strong opinion on the matter. In general, I tend to separate for simplicity and readability, but both options would work for me. On React Native, style organization can be a bit different than on the web, mainly due to the fact that RN styles can't be cascaded.

> fixed in [6778f08](https://github.com/Automattic/jetpack/commit/6778f08917e117cc89dbde6df309e9ccbf0e129f)

Great, thanks for updating this 🙇 .

File: projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx
Comment (fluiddot) [reply projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx]: I'd say so, at least I've seen this format in most of the RN files. Although in terms of performance, I don't think there's a real benefit. Some time ago, I recall it was encouraged to use styles created with `StyleSheet.create` and not modify them due to performance, but seems this is no longer the case ([StackOverflow discussion](https://stackoverflow.com/questions/38958888/react-native-what-is-the-benefit-of-using-stylesheet-vs-a-plain-object)).

**Review by fluiddot — COMMENTED** [review]
@jhnstn I noticed that one required job is failing (`
Linting / ESLint (non-excluded files only)`), looks like it's related to [this line](https://github.com/Automattic/jetpack/pull/29478/files#diff-40a6dfae85fa080df348681d0d9b92468585892fdc53a281898595833134122eR44). Once we address that, from my side we can consider this PR ready.

**UPDATE:** There's another required job that fails and would need to be addressed: `Test plugin upgrades`.

**Comment by fluiddot** [comment]
@jhnstn I see that ESLint is showing errors in the last commit ([CI job](https://github.com/Automattic/jetpack/actions/runs/4482717251/jobs/7881115399?pr=29478)):
```
jetpack/projects/packages/videopress/src/client/block-editor/blocks/video/components/player/index.native.tsx
Error:   43:22  error  Replace `(·<View·style={·style[·'videopress-player__overlay'·]·}·/>·)` with `<View·style={·style[·'videopress-player__overlay'·]·}·/>`  prettier/prettier

jetpack/projects/packages/videopress/src/client/block-editor/blocks/video/edit.native.tsx
Error:   21:1  error  `./components/playback-panel` import should occur before import of `./components/player`  import/order

**Review by SiobhyB — APPROVED** [review]

**Pull request merged**
